To access OpFastScintillation class from sbncode

marcodeltutto opened the pull request:
SBN needs to access `OpFastScintillation::detectedDirectHits()` and `OpFastScintillation::detectedReflecHits()` in order to get an estimation of the photons that arrive to optical detectors using the semi-analytical approach instead of the photon library. Currently, those two methods are private. This PR changes them to public. 

`detectedDirectHits` and `detectedReflecHits` should be provided by a service (similarly to the `PhotonVisibilityService`) and this is planned to be done in the future when things will be moved to `larg4`. For now, we need to access them in this way.

marcodeltutto committed:
Making detectedDirectHits and detectedReflecHits methods public.

marcodeltutto committed:
Use fPVS->NOpChannels() and not cached NOpChannels, so correct number of op channels is always returned

marcodeltutto committed:
Comment unused variable acos_arr

marcodeltutto committed:
Merge branch 'develop' into mdeltutt_opfast

FNALbuild commented:
A new Pull Request was created by  @marcodeltutto (Marco Del Tutto) for develop.

It involves the following packages:

larsim


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//732/console

absolution1 commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9714&builds= 
 +build

FNALbuild commented:
-dune tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/10369&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9713&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5618&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9713&builds=

knoepfel commented on larsim/LegacyLArG4/OpFastScintillation.hh:
```diff
@@ -464,8 +464,8 @@ namespace larg4 {
-  static std::array<double, acos_bins + 1>
-    acos_arr; // to get minimum resolution of 0.0000005 in [0,1]
+  // static std::array<double, acos_bins + 1>
+  //   acos_arr; // to get minimum resolution of 0.0000005 in [0,1]
```
Recommend removing the lines instead of commenting them out.

marcodeltutto replied:
Done with d829239.

knoepfel commented on larsim/LegacyLArG4/OpFastScintillation.cxx:
```diff
@@ -2192,15 +2192,15 @@ namespace larg4 {
+  //     exit(0);
+  //   }
+  //   return acos_arr[std::round(acos_bins * x)];
+  // }
```
Recommend removing instead of commenting out.

aihimmel replied:
Removing this was general cleanup? Not that it's a bad idea, just trying to understand since it seems unrelated to the main purpose of the PR.

marcodeltutto replied:
Lines removed with d829239.
It was cleanup but it was also related to this PR. The variable `acos_arr` was unused, causing `sbndcode` to fail compilation. I can of course change to compiler flags in `sbndcode` when it compiles this part, but I think it is cleaner to just remove the unused variable.

knoepfel commented on larsim/LegacyLArG4/OpFastScintillation.hh:
```diff
@@ -259,6 +259,13 @@ namespace larg4 {
+                            geo::Point_t const& ScintPoint);
+    void detectedReflecHits(std::map<size_t, int>& ReflDetectedNum,
+                            const double Num,
+                            geo::Point_t const& ScintPoint);
```
Now that these functions are public, is it possible to mark them as `const`?  This may require marking other member functions as `const` also.  The goal is to avoid adjusting the internal state of the class from outside.

marcodeltutto replied:
Yes, good point, thanks. Done with commit 8721ec9.

aihimmel commented on larsim/LegacyLArG4/OpFastScintillation.cxx:
```diff
@@ -1577,7 +1577,7 @@ namespace larg4 {
     // detemine hits on each PD
     const std::array<double, 3> hotspot = {plane_depth, ScintPoint.Y(), ScintPoint.Z()};
-    for (size_t const OpDet : util::counter(NOpChannels)) {
+    for (size_t const OpDet : util::counter(fPVS->NOpChannels())) {
```
Should the NOpChannels member be removed? Where else is it being used? If not, why not keep it here?

marcodeltutto replied:
Yes, thanks, done with a4c7c2e.

marcodeltutto replied:
It removed the `NOpChannels` member because it was not initialized in the constructor, and so depending on how you call this class you may not get `NOpChannels` properly initialized. I can either move the initialization to the constructor, or just remove this variable as the number of op dets is provided by the service. I preferred to go the second way as I thought it was cleaner.

aihimmel replied:
OK, sounds good good.

aihimmel commented on larsim/LegacyLArG4/OpFastScintillation.cxx:
```diff
@@ -1501,7 +1501,7 @@ namespace larg4 {
                                           geo::Point_t const& ScintPoint)
   {
-    for (size_t const OpDet : util::counter(NOpChannels)) {
+    for (size_t const OpDet : util::counter(fPVS->NOpChannels())) {
```
Should the NOpChannels member be removed? Where else is it being used? If not, why not keep it here?

marcodeltutto replied:
Yes, thanks, done with a4c7c2e.

marcodeltutto committed:
Marking detectedDirectHits and detectedReflecHits as const methods.

Also, mark as const all other methods required to be const to make this happen.

marcodeltutto committed:
Removing previously commented-out lines.

marcodeltutto committed:
Removing member NOpChannels. Always using fPVS->NOpChannels() to get the number of optical channels.

FNALbuild commented:
Pull request #44 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//741/console

marcodeltutto commented:
Hello, if possible, we would need this in the next LArSoft release. Please let me know if anything else is needed from my side. @knoepfel @aihimmel

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9786&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9785&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5033&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9785&builds=

FNALbuild commented:
-icarus tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/1219&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9785&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5229&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9785&builds=

FNALbuild commented:
-sbnd tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5598&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9785&builds=

FNALbuild commented:
-dune tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/10394&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9785&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5634&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9785&builds=

aihimmel commented:
Looks OK to me. Sorry for the delay. Approve.

wesketchum commented:
Thanks @aihimmel.

Regarding the failed/warning on the tests, probably an expert eye needed. It's not clear to me that these are related to this PR, or to other underlying issues common to other tests. Sorry if I'm adding noise, but just trying to help push this along.

lgarren commented:
@wesketchum  If you can help get those CI tests updated, we will be delighted. ;)

lgarren commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests.
